Add custom analytics documentation (#1487)
wso2/docs-mi: en/docs/mi-analytics/mi-custom-analytics.md

> # Publishing Custom Analytics Events Data

Reviewer: ```suggestion
# Publish Custom Analytics Events Data
```

We follow the imperative mood for titles. Please check other instances as well for consistency.

Author: Fixed with e2e4aaf143e54d7ca0242e669adee788bfff5196

## Introduction

WSO2 Micro Integrator provides the capability to publish custom analytics data with the existing event schema. This guide explains the steps required to achieve this.

This section outlines the process of creating a sample and configuring it with WSO2 Micro Integrator.

> ## Creating a sample

Reviewer: ```suggestion
## Create a sample
```

Author: Fixed with e2e4aaf143e54d7ca0242e669adee788bfff5196

To begin, create a new Java/Maven project.

!!! note
    Alternatively, you can use an existing [sample](https://github.com/wso2/product-micro-integrator/tree/master/samples/AnalyticsSampleDataProvider). If you choose to use the provided sample instead of developing one from scratch, you can skip the steps for creating the sample and proceed directly to [Configuring the Sample](#configuring-the-sample).

This section covers how to configure the `pom.xml` file, implement the required classes, and build the sample project.

> ### Configuring the pom.xml

Reviewer: ```suggestion
### Configure the pom.xml
```

Add the WSO2 Nexus repository to the `pom.xml` file:

```xml
    <repository>
        <id>wso2-nexus</id>
        <name>WSO2 internal Repository</name>
        <url>https://maven.wso2.org/nexus/content/groups/wso2-public/</url>
        <releases>
            <enabled>true</enabled>
            <updatePolicy>daily</updatePolicy>
            <checksumPolicy>ignore</checksumPolicy>
        </releases>
    </repository>
```

Add the following dependencies.

```xml
    <dependencies>
        <dependency>
            <groupId>org.wso2.ei</groupId>
            <artifactId>org.wso2.micro.integrator.analytics.messageflow.data.publisher</artifactId>
            <version>${wso2.mi.version}</version>
        </dependency>
        <dependency>
            <groupId>org.apache.synapse</groupId>
            <artifactId>synapse-extensions</artifactId>
            <version>${synapse.version}</version>
        </dependency>
        <dependency>
            <groupId>org.apache.synapse</groupId>
            <artifactId>synapse-core</artifactId>
            <version>${synapse.version}</version>
        </dependency>
    </dependencies>
```

!!! info

>     The versions for ${wso2.mi.version} and ${synapse.version} can be found in the jar versions available in the current Micro Integrator package.

Reviewer: ```suggestion
    The versions for ${wso2.mi.version} and ${synapse.version} can be found in the JAR versions available in the current Micro Integrator package.
```

Author: Fixed with 678ac7f19bf9c0d3fb87f3afe9416290c2189c99

> ### Implementing required class

Reviewer: ```suggestion
### Implement required class
```

Author: Fixed with e2e4aaf143e54d7ca0242e669adee788bfff5196

To add custom analytics data to the existing event schema, implement a class from the `org.wso2.micro.integrator.analytics.messageflow.data.publisher.producer.AnalyticsCustomDataProvider` interface. Override the `getCustomProperties` method to push custom analytics data.

An example implementation of this class is shown below:

``` java
    public class SampleCustomDataProvider implements AnalyticsCustomDataProvider {
        @Override
        public Map<String, Object> getCustomProperties(MessageContext messageContext) {
            Map<String, Object> properties = new HashMap<>();
            Mediator mediator = messageContext.getMainSequence();

>             properties.put("mediatorName", mediator.getMediatorName());

Reviewer 2: Shall we change this?

Author: Fixed with aa31633d778f7d581cfb691805ae25d7626da05b

            return properties;
        }
    }
```

### Build the project

To build the project, run the following command:

```
    mvn clean install
```

> ## Configuring the sample

Reviewer: ```suggestion
## Configure the sample
```

Author: Fixed with e2e4aaf143e54d7ca0242e669adee788bfff5196

This section outlines the steps required to configure WSO2 Micro Integrator for the sample created above.

> After building the project, copy the created .jar file into the `wso2mi-{version}/lib` directory.

Reviewer: ```suggestion
After building the project, copy the created `.jar` file into the `wso2mi-{version}/lib` directory.
```

> Edit the `analytics` configuration in the `deployment.toml` file, located inside the `wso2mi-{version}/conf` directory, and add the following configuration:

Reviewer: We usually follow the convention of using the `<MI_HOME>` placeholder instead of `wso2mi-{version}`.

Author: Fixed with e2e4aaf143e54d7ca0242e669adee788bfff5196

``` toml
    [analytics]
    enable = true
    custom_data_provider_class="org.wso2.ei.SampleCustomDataProvider"
```

This configuration ensures that the custom data provider class is engaged.

> Enable statistics for Micro Integrator following instructions [here](https://mi.docs.wso2.com/en/latest/mi-analytics/mi-elk-installation-guide/#enabling-statistics-for-artifacts).

Reviewer: ```suggestion
Enable statistics for Micro Integrator following the instructions [here]({{base_path}}/mi-analytics/mi-elk-installation-guide/#enabling-statistics-for-artifacts).
```

Author: Fixed with e2e4aaf143e54d7ca0242e669adee788bfff5196

> Create a log appender following instructions [here](https://mi.docs.wso2.com/en/latest/mi-analytics/mi-elk-installation-guide/#creating-log-appender).

Reviewer: ```suggestion
Create a log appender following the instructions [here]({{base_path}}/mi-analytics/mi-elk-installation-guide/#creating-log-appender).
```

Author: Fixed with e2e4aaf143e54d7ca0242e669adee788bfff5196

# Publishing Custom Analytics Events Data

## Introduction

WSO2 Micro Integrator provides the capability to publish custom analytics data with the existing event schema. This guide explains the steps required to achieve this.

This section outlines the process of creating a sample and configuring it with WSO2 Micro Integrator.

> ## Creating a Sample

Reviewer 2: ```suggestion
## Creating a sample
```

Author: Fixed with 951d0756979f1ac7962e630907200b4629e2a588

To begin, create a new Java/Maven project.

> Alternatively, you can use an existing [sample](https://github.com/wso2/product-micro-integrator/tree/master/samples/AnalyticsSampleDataProvider). If you choose to use the provided sample instead of developing one from scratch, you can skip the steps for creating the sample and proceed directly to [Configuring the Sample](#configuring-the-sample).

Reviewer 2: Let's add this as a note

Author: <img width="1002" alt="image" src="https://github.com/user-attachments/assets/8050220e-9b2e-4ff8-a7d2-ccace81d8ef0" />

Fixed with 951d0756

### Configuring the pom.xml

Add the WSO2 Nexus repository to the `pom.xml` file:

```xml
    <repository>
        <id>wso2-nexus</id>
        <name>WSO2 internal Repository</name>
        <url>https://maven.wso2.org/nexus/content/groups/wso2-public/</url>
        <releases>
            <enabled>true</enabled>
            <updatePolicy>daily</updatePolicy>
            <checksumPolicy>ignore</checksumPolicy>
        </releases>
    </repository>
```

Add the following dependencies.

```xml
    <dependencies>
        <dependency>
            <groupId>org.wso2.ei</groupId>
            <artifactId>org.wso2.micro.integrator.analytics.messageflow.data.publisher</artifactId>
            <version>${wso2.mi.version}</version>
        </dependency>
        <dependency>
            <groupId>org.apache.synapse</groupId>
            <artifactId>synapse-extensions</artifactId>
            <version>${synapse.version}</version>
        </dependency>
        <dependency>
            <groupId>org.apache.synapse</groupId>
            <artifactId>synapse-core</artifactId>
            <version>${synapse.version}</version>
        </dependency>
    </dependencies>
```

> **INFO:**
>

> > The versions for {wso2.mi.version} and ${synapse.version} can be found in the jar versions available in the current Micro Integrator package.

Author: Fixed with d1daa1edd873712e4c0932d29518800cd9014aa3

> 

> ### Implementing Required Class

Reviewer 2: ```suggestion
### Implementing required class
```

Author: Fixed with 951d0756979f1ac7962e630907200b4629e2a588

> #### AnalyticsCustomDataProvider Implementation Class

Reviewer 2: ```suggestion
#### AnalyticsCustomDataProvider implementation class
```

Author: Fixed with 951d0756979f1ac7962e630907200b4629e2a588
